### ISSUE TITLE
Non sfc function shows "react/no-this-in-sfc"

### ISSUE THREAD

**Issue by jmdsg** [yannickcr/eslint-plugin-react#2010]
Hi, i have the following file which does not have a stateless functional component, it does not involve react at all but is showing me the error `react/no-this-in-sfc` when i call `this.connection`.

```javascript
import { ValidatedMethod } from "meteor/mdg:validated-method";
import { SimpleSchema } from "meteor/aldeed:simple-schema";

export const prepareLogin = new ValidatedMethod({
    name: "user.prepare",
    validate: new SimpleSchema({
        // ...
    }).validator(),
    run({ remember }) {
        if (Meteor.isServer) {
            const connectionId = this.connection.id; // react/no-this-in-sfc
            return Methods.prepareLogin(connectionId, remember);
        }
        return null;
    },
});
```

**Comment by ljharb** [comment]
cc @alexzherdev any chance this is fixed already on master?

**Comment by alexzherdev** [comment]
No, it's not. This is not in a class, nor in a class method, so our checks are not triggering here. Perhaps we might add a check for a member in an object expression? But then what about this valid, albeit likely rare, use case 😕 
```jsx
const Namespace = {
  Component(props) {
    return <div />;
  }
};

<Namespace.Component />
```

**Comment by ljharb** [comment]
In this case, we’re determining it’s an SFC because it’s a function that destructured it’s first parameter, and sometimes returning null - but I’d expect that being a property of an object - especially if it’s an object passed to a constructor or a function - should downgrade its confidence so that it doesn’t report. Additionally, the function’s name is not PascalCased.

**Comment by ElvenMonky** [comment]
Please, take into consideration, that SFC doesn't have to (or even shouldn't) have name at all.

**Comment by ljharb** [comment]
Every React component absolutely *should* have a name, for debugging purposes. No component (SFC or not) *has* to.

**Comment by ElvenMonky** [comment]
Debugging is not an excuse to pollute your code by dublicating indefiers. If one desires displayName, it can be attached explicitely using "displayName" property and in separate debugging module, rather than relying on React implicitely and mixing debug and production code in same modules. Same stands for prop-types.

**Comment by ElvenMonky** [comment]
Sorry, but I don't see why my comment was marked an off-topic.
Once again, most SFC are arrow functions, which can't have names. So if you'd like to take function name into consideration for "react/no-this-in-sfc" - do it with caution.

**Comment by ljharb** [comment]
Most SFCs aren’t necessarily arrow functions, because the Airbnb style guide forbids them - either way, the absence of a name surely won’t mean it’s not an SFC, but the presence of one might.

**Comment by AnabelSalomone** [comment]
I've seen a fix for an issue like this one was made (here: https://github.com/yannickcr/eslint-plugin-react/pull/1995 ), any chance we can have it in a release soon?

**Comment by ljharb** [comment]
[v7.12.0](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.12.0) is released. Is this issue still a problem?

**Comment by psachs21** [comment]
@ljharb i have confirmed that this is still a problem as of [v7.14.3](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.14.3) Going to checkout the code and see why exactly it's identifying my example as a functional component.

```js
@connect(state => ({
  someReduxState: state.someReduxState
}))
export default class GroupWebUserWizard extends BaseComponent {
  renderFooter() {
    return <div>{this.formatMessage(...)}</div>
  }
}
```

**Comment by psachs21** [comment]
I realize my mistake, my example looks like this:

```js
export default class {
  renderFooter() {
    return () => (
      <div>{this.value}</div>
    );
  }
}
```

Which is the thing that threw me.

**Comment by ljharb** [comment]
Your arrow function there is considered an SFC.

**Comment by psachs21** [comment]
I thought so as well, but apparently this passes just fine:

```js
export default class {
  renderFooter = () => () => (
      <div>{this.value}</div>
    );
  }
}
```

But this doesn't:

```js
export default class {
  renderFooter = () => {
    return () => (
      <div>{this.value}</div>
    );
    }
  }
}
```

**Comment by ljharb** [comment]
interesting, I’d expect both, or neither, to be detected as a component.

**Comment by isachivka** [comment]
```
export function wrapWithContext(context, contextTypes, children) {
  class ContextWrapper extends Component {
    static childContextTypes = contextTypes;

    constructor(props) {
      super(props);
      // eslint-disable-next-line react/no-this-in-sfc
      this.state = {};
    }

    getChildContext() {
      return context;
    }

    setProps = (props) => {
      // eslint-disable-next-line react/no-this-in-sfc
      this.setState({ ...props });
    };

    render() {
      return (
        // eslint-disable-next-line react/no-this-in-sfc
        cloneElement(children, this.state)
      );
    }
  }
  return createElement(ContextWrapper);
}
```

**Comment by ljharb** [comment]
No idea why setProps is an arrow function there (functions should never go in class fields) but certainly that's a clear example of a bug in the rule, thanks!
